skimage/_shared/utils.py

```diff
@@ -450,3 +450,31 @@ def _fix_ndimage_mode(mode):
     if NumpyVersion(scipy.__version__) >= '1.6.0':
         mode = grid_modes.get(mode, mode)
     return mode
+
+
+def _float_type(image):
+    """Return an appropriate floating-point dtype for a given image.
+
+    float32, float64, complex64, complex128 are preserved.
+    float16 is promoted to float32.
+    complex256 is demoted to complex128.
+    Other types are cast to float64.
+
+    Paramters
+    ---------
+    image : ndarray
+        Input image.
+
+    Retruns
+    -------
+    float_type : dtype
+        Floating-point dtype for the image.
+    """
+    char = image.dtype.char
+    if char == 'e':
+        return np.float32
+    elif char == 'G':
+        return np.complex128
+    elif char not in 'dfDF':
+        return np.float64
+    return image.dtype
```

Comment:
Also, other types should be cast to `np.float64`; either the docstring needs an update or the last line does to make it consistent.

Reply:
see "char not in" two lines above. Only the floating point types we want to keep ('dfDF') can reach that last line. We don't want to cast complex64 to float64 for instance.

We could potentially add an allow_complex flag to this utility that would just raise a ValueError on complex dtypes since the majority of functions in scikit-image deal with real-valued inputs.

Reply:
ah right. I missed that! my bad. As I understand you are happy with the dict; otherwise I would suggest switching the logic of the last `elif` and the final return statement to make the final return the catch-all/default statement. Just to create a more intuitive `if ... elif ... else` style.

Not sure about the flag idea. Do you think adding complexity here will make our life easier somewhere else?